driver/src/main/java/org/kaazing/k3po/driver/internal/netty/bootstrap/file/FileChannelSink.java

```diff
@@ -0,0 +1,181 @@
+import java.io.RandomAccessFile;
+import java.net.URI;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel.MapMode;
+import java.util.Map;
+
+import static java.nio.channels.FileChannel.MapMode.READ_ONLY;
+import static java.nio.channels.FileChannel.MapMode.READ_WRITE;
+import static org.jboss.netty.channel.Channels.fireChannelBound;
+
+public class FileChannelSink extends AbstractChannelSink {
+
+    private static final InternalLogger LOGGER = InternalLoggerFactory.getInstance(FileChannelSink.class);
+
+    // Note that read position and write position are independent in the script. We are keeping
+    // two buffers(view on the same MappedByteBuffer) to achieve this and also don't have to
+    // manage positions explicitly
+    private ChannelBuffer readBuffer;
+    private ChannelBuffer writeBuffer;
+
+    @Override
+    protected void connectRequested(ChannelPipeline pipeline, ChannelStateEvent evt) throws Exception {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("connectRequested pipeline = " + pipeline + " evt = " + evt);
+        }
+
+        ChannelFuture connectFuture = evt.getFuture();
+        ChannelAddress fileAddress = (ChannelAddress) evt.getValue();
+        Map<String, Object> options = fileAddress.getOptions();
+        String mode = (String) options.get("mode");
+        if (mode == null) {
+            mode = "rw";
+        }
+        Long size = (Long) options.get("size");
+        if (size == null) {
+            size = (long) 0;
+        }
+        FileChannel fileChannel = (FileChannel) evt.getChannel();
+
+        try {
+            MappedByteBuffer mappedBuffer = mapFile(fileAddress.getLocation(), mode, size);
+            readBuffer = ChannelBuffers.wrappedBuffer(mappedBuffer);
+            setReadOffset(0);
+            writeBuffer = ChannelBuffers.wrappedBuffer(mappedBuffer);
+            setWriteOffset(0);
+
+            if (!fileChannel.isBound()) {
+                fileChannel.setLocalAddress(fileAddress);
+                fileChannel.setBound();
+                fireChannelBound(fileChannel, fileAddress);
+            }
+
+            connectFuture.setSuccess();
+        } catch (Throwable t) {
+            connectFuture.setFailure(t);
+        }
+
+        Channels.fireChannelConnected(fileChannel, fileAddress);
+
+        // Send a read event using memory mapped buffer contents so that reads in the
+        // scripts can be matched
+        fireMessageReceived(fileChannel, fileAddress);
+    }
+
+    @Override
+    protected void writeRequested(ChannelPipeline pipeline, MessageEvent evt) throws Exception {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("writeRequested pipeline = " + pipeline + " evt = " + evt);
+        }
+        ChannelBuffer channelBuffer = (ChannelBuffer) evt.getMessage();
+        writeBuffer.writeBytes(channelBuffer);
+        ChannelFuture writeFuture = evt.getFuture();
+        writeFuture.setSuccess();
```

Comment:
What happens if there is insufficient space at the end of the file to accommodate the write request?